core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -65,7 +65,7 @@ private[spark] case class HeartbeatResponse(reregisterBlockManager: Boolean)
  * Lives in the driver to receive heartbeats from executors..
  */
 private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
-  extends SparkListener with ThreadSafeRpcEndpoint with Logging {
+  extends SparkListener with ThreadSafeRpcEndpoint with IsolatedRpcEndpoint with Logging {
```

Comment:
Could you elaborate your scale of testing? Are you using this in the production?

Reply:
Do we need both `ThreadSafeRpcEndpoint` or `IsolatedRpcEndpoint`?

Reply:
> Do we need both `ThreadSafeRpcEndpoint` or `IsolatedRpcEndpoint`?

In theory, we only needs `IsolatedRpcEndpoint` as the default thread is 1, but I'm not sure whether there're some logic depending on trait `ThreadSafeRpcEndpoint`. To keep it safe, I just added new trait

Reply:
> Could you elaborate your scale of testing? Are you using this in the production?

App with about 500+ executors. Yes, it's for production.

Reply:
I understand what you mean, but it's still not-safe completely, isn't it? If you are considering two different logic branches based on `ThreadSafeRpcEndpoint` or `IsolatedRpcEndpoint`, this PR could break some of the existing logic if the branching logic starts to match `IsolatedRpcEndpoint` first.
> I'm not sure whether there're some logic depending on trait ThreadSafeRpcEndpoint. To keep it safe, I just added new trait

Reply:
Agree with @dongjoon-hyun - we dont need both.
Given number of threads == 1, this should functionally behave like it is thread safe though.

I think the change of replace `ThreadSafeRpcEndpoint` with `IsolatedRpcEndpoint` should be fine.
But would be great if you can take a look @Ngone51.
Also, +CC @JoshRosen

Reply:
+1 for @mridulm 's comment.